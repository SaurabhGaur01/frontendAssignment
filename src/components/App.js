import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'redux';
import TableComponent from './Table/Table';
import Chart from './Chart/Chart';
import TabPanel from './Tabs/TabPanel';
import TabHeader from './Tabs/TabHeader';
import handleRetrieveData from '../thunks/handleRetrieveData';

const App = ({ actionHandleRetrieveData }) => {
    /* 
    Below action will be dispatched before component renders and make an API call,
    store the response in redux store
    */
    React.useEffect(() => {
        actionHandleRetrieveData();
    }, []);

    // Using hook for tab switch purpose
    const [tabValue, setTabValue] = React.useState(0);
    return (
        <React.Fragment>
            <TabHeader value={tabValue} handleChange={(event, newValue)=>{setTabValue(newValue)}} />
            <TabPanel value={tabValue} index={0}>
                <TableComponent />
            </TabPanel>
            <TabPanel value={tabValue} index={1}>
                <Chart />
            </TabPanel>
        </React.Fragment>       
    );
}

App.propTypes = {
    actionHandleRetrieveData: PropTypes.func.isRequired,
};

// Getting action from redux store for dispatching 
const mapDispatchAsProps = {
    actionHandleRetrieveData: handleRetrieveData,
}

const hocChain = compose(
    connect(null, mapDispatchAsProps),
);

export { App as TestableApp };

export default hocChain(App);