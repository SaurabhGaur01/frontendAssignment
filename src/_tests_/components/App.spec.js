import React from 'react';
import { shallow } from 'enzyme';
import { TestableApp } from '../../components/App';

describe('App Component', () => {
    let props;

    beforeEach(() =>{
        props = {
            actionHandleRetrieveData: jest.fn(),
        };
    });

    it('should render correctly', () => {
        const renderedModule = shallow(<TestableApp {...props} />);
        expect(renderedModule).toMatchSnapshot();
    });

    it('should call actionHandleRetrieveData when component mounts', () => {
        jest.spyOn(React, 'useEffect').mockImplementationOnce(f => f());
        shallow(<TestableApp {...props} />);
        expect(props.actionHandleRetrieveData).toHaveBeenCalledTimes(1);
    });
})