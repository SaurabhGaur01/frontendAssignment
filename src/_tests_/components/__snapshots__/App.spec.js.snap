// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should render correctly 1`] = `
<Fragment>
  <TabHeader
    handleChange={[Function]}
    value={0}
  />
  <TabPanel
    index={0}
    value={0}
  >
    <Connect(TableComponent) />
  </TabPanel>
  <TabPanel
    index={1}
    value={0}
  >
    <Connect(Chart) />
  </TabPanel>
</Fragment>
`;
